// @affinidi/passport-affinidi/dist/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.affinidiProvider = void 0;
const tslib_1 = require("tslib");
const express_session_1 = tslib_1.__importDefault(require("express-session"));
const openid_client_1 = require("openid-client");
const passport_1 = tslib_1.__importDefault(require("passport"));
const profile_1 = require("./profile");
const strategy_1 = tslib_1.__importDefault(require("./strategy"));
const affinidiProvider = async (app, options) => {
    const { client, strategy, sessionKey } = await (0, strategy_1.default)(options);
    passport_1.default.use('affinidi-oidc', strategy);
    // app.use(passport.initialize());
    // app.use(passport.session());
    app.use((0, express_session_1.default)({
        secret: options.id,
        resave: false,
        saveUninitialized: true,
        cookie: {
            secure: process.env.NODE_ENV === 'production',
            maxAge: 1000 * 60 * 60 * 24 * 1, // 1 day
        },
        unset: 'destroy',
        ...options.expressSesssion,
    }));
    // handles serialization and deserialization of authenticated user
    // passport.serializeUser(function (user, done) {
    //     done(null, user);
    // });
    // passport.deserializeUser(function (user, done) {
    //     done(null, user);
    // });
    const initHandler = (req, res, next) => {
        const code_verifier = openid_client_1.generators.codeVerifier();
        const params = {
            code_challenge: openid_client_1.generators.codeChallenge(code_verifier),
            code_challenge_method: 'S256',
            response_type: 'code',
            scope: 'openid',
            state: openid_client_1.generators.state(),
        };
        req.session[sessionKey] = {
            state: params.state,
            response_type: params.response_type,
            code_verifier,
        };
        const authorizationUrl = client.authorizationUrl(params);
        res.send({ authorizationUrl });
    };
    const completeHandler = (req, res, next) => {
        passport_1.default.authenticate('affinidi-oidc', {}, function (err, user, info) {
            if (err) {
                if (options.onError && typeof options.onError === 'function') {
                    options.onError(err, info);
                }
                res.status(400).send({
                    error: err.message,
                    error_description: err.error_description,
                });
            }
            else {
                const profile = (options.profileParser && options.profileParser(user)) || (0, profile_1.profileParser)(user);
                if (options.onSuccess && typeof options.onSuccess === 'function') {
                    options.onSuccess(user, profile, info);
                }
                res.send({ user: profile });
            }
        })(req, res, next);
    };
    app.get(options.routes?.init || '/api/affinidi-auth/init', options.routes?.initHandler || initHandler);
    app.post(options.routes?.complete || '/api/affinidi-auth/complete', options.routes?.completeHandler || completeHandler);
};
exports.affinidiProvider = affinidiProvider;
